pybedlite/overlap_detector.py

```diff
@@ -126,36 +177,94 @@ def from_bedrecord(cls: Type["Interval"], record: BedRecord) -> "Interval":
+      * `strand (BedStrand)`: The BED strand of the feature
+      * `strand (str)`: The strand of the feature (`"-"` for negative)
 
     The same interval may be added multiple times, but only a single instance will be returned
-    when querying for overlaps.  Intervals with the same coordinates but different names are
-    treated as different intervals.
+    when querying for overlaps.
 
     This detector is the most efficient when all intervals are added ahead of time.
     """
 
-    def __init__(self) -> None:
+    def __init__(self, intervals: Optional[Iterable[_GenericGenomicSpan]] = None) -> None:
         # A mapping from the contig/chromosome name to the associated interval tree
         self._refname_to_tree: Dict[str, cr.cgranges] = {}  # type: ignore
         self._refname_to_indexed: Dict[str, bool] = {}
-        self._refname_to_intervals: Dict[str, List[Interval]] = {}
+        self._refname_to_intervals: Dict[str, List[_GenericGenomicSpan]] = {}
+        if intervals is not None:
+            self.add_all(intervals)
 
-    def __iter__(self) -> Iterator[Interval]:
+    def __iter__(self) -> Iterator[_GenericGenomicSpan]:
         """Iterates over the intervals in the overlap detector."""
         return itertools.chain(*self._refname_to_intervals.values())
 
-    def add(self, interval: Interval) -> None:
+    @staticmethod
+    def _reference_sequence_name(interval: GenomicSpan) -> str:
+        """Return the reference name of a given interval."""
+        if isinstance(interval, Interval) or hasattr(interval, "refname"):
+            return interval.refname
+        elif isinstance(interval, BedRecord) or hasattr(interval, "chrom"):
+            return interval.chrom
+        elif hasattr(interval, "contig"):
+            return interval.contig
+        else:
+            raise ValueError(
+                f"Genomic span is missing a reference sequence name property: {interval}"
+            )
+
+    @staticmethod
+    def _is_negative(interval: GenomicSpan) -> bool:
+        """Determine if this is a negative stranded interval or not."""
+        return (
+            (hasattr(interval, "negative") and interval.negative)
+            or (
+                hasattr(interval, "strand")
+                and isinstance(interval.strand, BedStrand)
+                and interval.strand is BedStrand.Negative
+            )
+            or (
+                hasattr(interval, "strand")
+                and isinstance(interval.strand, str)
+                and interval.strand == "-"
+            )
+        )
+
+    def add(self, interval: _GenericGenomicSpan) -> None:
         """Adds an interval to this detector.
 
         Args:
             interval: the interval to add to this detector
         """
-        refname = interval.refname
+        if not isinstance(interval, Hashable):
+            raise ValueError(f"Interval feature is not hashable but should be: {interval}")
```

Comment:
Small feature addition that would be nice to have regardless of this PR.

This is necessary because in some class methods, `set` is called.

Reply:
As not to pay the performance here, can we move this to where `set` is called and wrap it with a `try`/`except` for a `TypeError` that provides this exception message?  I think that's here:

https://github.com/fulcrumgenomics/pybedlite/blob/74f7d031f48036fd2aa2646b17d4f375669e36fe/pybedlite/overlap_detector.py#L366-L368

Reply:
Also, since `GenomicSpan` sub-classes `Hashable`, is this check redundant?

Reply:
> Also, since `GenomicSpan` sub-classes `Hashable`, is this check redundant?

TLDR: with this, we have 1 check at the type-level and 1 check at the implementation-level; both serve a purpose

Having `GenomicSpan` subclass `Hashable` ensures _**type safety**_, so MyPy will catch those misuses.

However, MyPy doesn't run at the _**implementation**_-level so that's where this check come in.

I do think both are important for a good user experience.

---

> As not to pay the performance here, can we move this to where `set` is called and wrap it with a `try`/`except` for a `TypeError` that provides this exception message? I think that's here:
> 
> https://github.com/fulcrumgenomics/pybedlite/blob/74f7d031f48036fd2aa2646b17d4f375669e36fe/pybedlite/overlap_detector.py#L366-L368

This is a good idea If we commonly **_store_** many more features than we return by the methods of this class. This is likely to be the common use case, so I'm happy to make this change!